xfrd.c

```diff
@@ -2377,77 +2511,106 @@ find_same_master_notify(xfrd_zone_type* zone, int acl_num_nfy)
+		if(!zone->latest_xfr)
+			continue;
+		xfr = zone->latest_xfr;
+		while(!xfr->sent && xfr->prev) {
+			xfr = xfr->prev;
+		}
+
+		/* zone has sent update and no (or different) nsd soa, the
+		   update must be corrupt */
+		if(xfr->sent == sent &&
 			(zone->soa_nsd_acquired == 0 ||
-			zone->soa_disk.serial != zone->soa_nsd.serial))
+			 zone->soa_nsd.serial != htonl(xfr->msg_new_serial)))
 		{
-			if(zone->soa_disk_acquired <
-				xfrd->reload_cmd_last_sent)
-			{
-				/* this zone should have been loaded, since its disk
-				   soa time is before the time of the reload cmd. */
-				xfrd_soa_type dumped_soa = zone->soa_disk;
-				log_msg(LOG_ERR, "xfrd: zone %s: soa serial %u "
-						 		 "update failed, restarting "
-						 		 "transfer (notified zone)",
-					zone->apex_str, (unsigned)ntohl(zone->soa_disk.serial));
-				/* revert the soa; it has not been acquired properly */
-				if(zone->soa_disk_acquired == zone->soa_nsd_acquired) {
-					/* this was the same as served,
-					 * perform force_axfr , re-download
-					 * same serial from master */
-					zone->soa_disk_acquired = 0;
-					zone->soa_nsd_acquired = 0;
-				} else {
-					/* revert soa to the one in server */
-					zone->soa_disk_acquired = zone->soa_nsd_acquired;
-					zone->soa_disk = zone->soa_nsd;
-				}
-				/* pretend we are notified with disk soa.
-				   This will cause a refetch of the data, and reload. */
-				xfrd_handle_incoming_notify(zone, &dumped_soa);
-				xfrd_set_timer_refresh(zone);
-			} else if(zone->soa_disk_acquired >= xfrd->reload_cmd_last_sent) {
-				/* this zone still has to be loaded,
-				   make sure reload is set to be sent. */
-				if(xfrd->need_to_send_reload == 0 &&
-					xfrd->reload_added == 0) {
-					log_msg(LOG_ERR, "xfrd: zone %s: needs "
-									 "to be loaded. reload lost? "
-									 "try again", zone->apex_str);
-					xfrd_set_reload_timeout();
-				}
+			xfrd_soa_type soa;
+			soa.serial = htonl(xfr->msg_new_serial);
+			log_msg(LOG_ERR, "xfrd: zone %s: soa serial %u update "
+			                 "failed, restarting transfer "
+			                 "(notified zone)",
+			                 zone->apex_str, xfr->msg_new_serial);
+			/* revert the soa; it has not been acquired properly */
+			if(xfr->acquired == zone->soa_nsd_acquired) {
+				/* this was the same as served,
+				 * perform force_axfr , re-download
+				 * same serial from master */
+				zone->soa_disk_acquired = 0;
+				zone->soa_nsd_acquired = 0;
+			} else {
+				/* revert soa to the one in server */
+				zone->soa_disk_acquired = zone->soa_nsd_acquired;
+				zone->soa_disk = zone->soa_nsd;
+			}
+			/* fabricate soa and trigger notify to refetch and
+			 * reload update */
+			memset(&soa, 0, sizeof(soa));
+			soa.serial = htonl(xfr->msg_new_serial);
+			xfrd_handle_incoming_notify(zone, &soa);
+			xfrd_set_timer_refresh(zone);
+			/* delete all pending updates */
+			xfr = zone->latest_xfr;
+			while(xfr) {
+				if(!xfr->acquired)
+					continue;
```

Comment:
This is an infinite loop if !xfr->acquired, it needs to go to the xfr->prev before continue.

Reply:
(l)oops (sorry, couldn't resist :sweat_smile:, thanks, will fix)

Reply:
done